flink-ml-lib/src/main/java/org/apache/flink/ml/classification/knn/Knn.java

```diff
@@ -0,0 +1,155 @@
+                            @Override
+                            public void mapPartition(
+                                    Iterable<Tuple3<DenseVector, Double, Double>> values,
+                                    Collector<KnnModelData> out) {
+                                List<Tuple3<DenseVector, Double, Double>> buffer =
+                                        new ArrayList<>(1);
+                                for (Tuple3<DenseVector, Double, Double> value : values) {
+                                    buffer.add(value);
+                                }
+                                int featureDim = buffer.get(0).f0.size();
+                                DenseMatrix packedFeatures =
+                                        new DenseMatrix(featureDim, buffer.size());
+                                DenseVector featureNorms = new DenseVector(buffer.size());
+                                DenseVector labels = new DenseVector(buffer.size());
+                                int offset = 0;
+                                for (Tuple3<DenseVector, Double, Double> data : buffer) {
+                                    System.arraycopy(
+                                            data.f0.values,
+                                            0,
+                                            packedFeatures.values,
+                                            offset * featureDim,
+                                            featureDim);
+                                    labels.values[offset] = data.f1;
+                                    featureNorms.values[offset++] = data.f2;
+                                }
+                                out.collect(new KnnModelData(packedFeatures, featureNorms, labels));
+                            }
+                        });
+        modelData.getTransformation().setParallelism(1);
+        return modelData;
+    }
+
+    /**
+     * For Euclidean distance, distance = sqrt((a - b)^2) = (sqrt(a^2 + b^2 - 2ab)) So it can
+     * pre-calculate the L2 norm square of the feature vector, and when calculating the distance
+     * with another feature vector, only dot product is calculated.
+     *
+     * @param inputData Input data.
+     * @return Input data with norm.
+     */
+    DataStream<Tuple3<DenseVector, Double, Double>> computeNorm(DataStream<Row> inputData) {
+        return inputData.map(
+                new MapFunction<Row, Tuple3<DenseVector, Double, Double>>() {
+                    @Override
+                    public Tuple3<DenseVector, Double, Double> map(Row value) {
+                        Double label = (Double) value.getField(getLabelCol());
+                        DenseVector feature = (DenseVector) value.getField(getFeaturesCol());
+                        double featureNorm = 0;
+                        for (int i = 0; i < feature.size(); ++i) {
+                            featureNorm += feature.values[i] * feature.values[i];
```

Comment:
Is this L2 norm? L2 norm is \sqrt(_sum_i x_i^2), right?

Reply:
Yes, I will rename it later.